WireDomain/Sources/WireDomain/Repositories/Client/ClientRepository.swift

```diff
@@ -0,0 +1,199 @@
+/// Facilitate access to clients related domain objects.
+///
+/// A repository provides an abstraction for the access and storage
+/// of domain models, concealing how and where the models are stored
+/// as well as the possible source(s) of the models.
+public protocol ClientRepositoryProtocol {
+
+    /// Fetches self user clients.
+    /// - returns : A self user clients list.
+
+    func fetchSelfClients() async throws -> [WireAPI.UserClient]
+
+    /// Fetches user clients.
+    /// - parameter userIDs: A list of user qualified ids.
+    /// - returns : A list of clients for a given user on a given domain.
+
+    func fetchClients(for userIDs: Set<UserID>) async throws -> [WireAPI.UserClients]
+
+    /// Fetches or creates a client locally.
+    ///
+    /// - parameters:
+    ///     - id: The user client id to find or create locally.
+    /// - returns: The user client found or created locally and a flag indicating whether or not the user client is new.
+
+    func fetchOrCreateClient(
+        with id: String
+    ) async throws -> (client: WireDataModel.UserClient, isNew: Bool)
+
+    /// Updates the user client informations locally.
+    ///
+    /// - parameters:
+    ///     - id: The user client id.
+    ///     - remoteClient: The up-to-date remote user client.
+    ///     - isNewClient: A flag indicating whether the user client is new.
+
+    func updateClient(
+        with id: String,
+        from remoteClient: WireAPI.UserClient,
+        isNewClient: Bool
+    ) async throws
+
+    /// Deletes client locally.
+    /// - parameter id: The client id.
+
+    func deleteClient(with id: String) async
+}
+
+public struct ClientRepository: ClientRepositoryProtocol {
+
+    // MARK: - Properties
+
+    private let clientAPI: any ClientAPI
+    private let context: NSManagedObjectContext
+
+    // MARK: - Object lifecycle
+
+    init(
+        clientAPI: any ClientAPI,
+        context: NSManagedObjectContext
+    ) {
+        self.clientAPI = clientAPI
+        self.context = context
+    }
+
+    // MARK: - Public
+
+    public func fetchSelfClients() async throws -> [WireAPI.UserClient] {
+        try await clientAPI.getSelfClients()
+    }
+
+    public func fetchClients(for userIDs: Set<UserID>) async throws -> [WireAPI.UserClients] {
+        try await clientAPI.getClients(for: userIDs)
+    }
+
+    public func fetchOrCreateClient(
+        with id: String
+    ) async throws -> (client: WireDataModel.UserClient, isNew: Bool) {
+        let localUserClient = await context.perform { [context] in
+            if let existingClient = UserClient.fetchExistingUserClient(
+                with: id,
+                in: context
+            ) {
+                return (existingClient, false)
+            } else {
+                let newClient = UserClient.insertNewObject(in: context)
+                newClient.remoteIdentifier = id
+                return (newClient, true)
+            }
+        }
+
+        try context.save()
```

Comment:
```suggestion
        if !localUserClient.1 {
           try context.save()
        }
```
we only need to save when the client is new

Reply:
I actually removed it in https://github.com/wireapp/wire-ios/pull/2019/commits/de1f3cf2daf03969a031a3d5fc73a19402d26eb4 because we are discussing about our saving strategy in another PR and I think we will need to be aligned about that